### ISSUE TITLE
Troubles getting add-in updates

### ISSUE THREAD

**Issue by AlexPhilippov** [OfficeDev/office-js#1536]
Our users having troubles getting the last updated version of the add-in: https://appsource.microsoft.com/en-us/product/office/WA104379640
1. User doesn't get the last version automatically. It's required to uninstall the add-in and then install it again to get the latest changes. Update panel appears for some users (mostly MacOS) but it seems to happen randomly, we can't rely on that. The only way for a common user to get the latest changes is the add-in reinstallation.
2. Removing add-in and then installing it from Web-store doesn't register custom functions, #NAME appears as a result. Sometimes, an "Error installing functions" message appears.
Issues reported on each version update. On version 3.3.0.0, we switched to Shared Runtime for custom functions. It doesn't improve the situation.

## Expected Behavior
1. User will see the update panel after the new version is approved.
2. Reinstallation from Web-store will provide a fully-functioning add-in with custom functions.

## Current Behavior
1. Panel update doesn't appear instantly after version approval. Some users won't see it at all, some will see it the next day.
2. Reinstallation from Web-store installs add-in without custom functions.

## Steps to Reproduce, or Live Example
0. Updated office, windows, and restarted computer. 
1. In Excel, remove the previously-installed add-in by going to Data > right click on DataFinder and Remove Add-in. Close Excel.
2. Open Chrome. Clear cache. Go to https://knoema.com/datafinder > Microsoft Appsource > login > Click get it now. Click Continue. Click Open in Excel.
3. Click Data Finder in Data tab. Click Insert Data for production of lemons in Italy.
Values will show up as #NAME? and Excel formulas (f.e. = KNOEMA.VERSION()) will not work either

## Context
We need to have a reliable way to provide version updates to our users.

## Your Environment
PC, Excel, Office 365 MSO (16.0.13426.20308) 64-bit, Windows 10.0.19041

## Useful logs
Attached OfficeAddins log file for case #2. There are no functions with the namespace "KNOEMA" loaded. (KNOEMA_DEV is a developers version with a different Id, shouldn't affect the production version)
[OfficeAddins.log - Copy.txt](https://github.com/OfficeDev/office-js/files/5641797/OfficeAddins.log.-.Copy.txt)

**Comment by xiaochun** [comment]
@AlexPhilippov Thanks to reach us. Sorry to hear that the issue bothers you for a while. I am trying the same locally with repro steps and noticed the java script file for custom functions may be missed to be referenced in "_**Knoema.DataFinder.Url**_ (https://addins.knoema.com/KnoemaDataFinder/Start.html)". You may can add `<script type="text/javascript" src="https://addins.knoema.com/KnoemaDataFinder/3.3.0.2/Js/functions.js" ></script>` in **_Knoema.DataFinder.Url_** to have a try (it can be placed just after "jquery" one.). Please let me know if it works or not.

**Comment by AlexPhilippov** [comment]
@xiaochun, thanks, that helped with the 2nd point. Any information on how can we ensure that clients will get the latest available version after the update(point 1)?

**Comment by grangeryy** [comment]
@AlexPhilippov Great to know it works for #2 problem. For #1 issue, we are still contacting the partner team on the updating process and will let you know once we have more information.

**Comment by jargil** [comment]
@AlexPhilippov - thanks for reporting your concerns.

I looked at the updates situation and I was not able to reproduce any problems with it locally. One thing to clarify is that updates don't happen immediately after they have been approved. It can take from a few hours to a few days for different computers. There are other internal factors that could also add delays in updating an add-in on a given computer.

If you've noticed users that haven't received updates after several days, please ask them to send us feedback via a Frown directly in their Office app. We'll be able to receive information that will help us identify and investigate any possible problems with Office.

**Comment by AlexPhilippov** [comment]
@jargil, thanks for the response.
It's quite unfortunate that we need to wait a few days to get an updated version. Along with the approval process, it's taking now somewhat about a week to provide a new version to a common user.
Are there any known factors that delay the updates? Is there any reliable way for common users to update their version to the latest except for add-in reinstallation and clearing Wef-folder?

**Comment by lindalu-MSFT** [comment]
Reopening due to ongoing communication.

**Comment by jargil** [comment]
It's all timing and expiration of cached items that drive our internal update logic. My recommendation is to wait for Office's internal logic to kick in and do its updates rather than trying to force it. Obviously this discussion is only relevant for manifest updates. You can always make logic changes to your add-in's source code and those will update independently of how often Office gets a manifest update. Code changes follow regular Web/Browser Standards for when they get updated.

@AlexPhilippov - I'm interested in understanding how often you find yourself having to make manifest updates vs logic/code changes?

Thanks for your feedback.

**Comment by AlexPhilippov** [comment]
@jargil, I can give you the whole story of why we're trying manifest updates now to get the latest version.
1. We're started with a simple static application stored on Amazon S3 + CloudFront that works only inside the add-in panel. We have a starting html-file that redirects to the latest version and correct office-host content since we need different implementations for Excel/Word. We updated this starting file to redirect to a new version, invalidated it within CDN, and the panel was updated quite reliably without any manifest updates.
2. Then we added custom functions as a separate runtime. We need consistency between add-in panel content and custom-functions version. And expiration of the custom-functions file wasn't the same as starting html-file. We could get a new html-file version while functions weren't updated. So we started to perform updates via submitting new manifests after each custom-functions file changes. And that takes a lot of time.
3. So now we added shared runtime for manifest expecting that custom-functions will be updated the same way it's done for other sources from html. But manifest-file specifies the link to the custom-functions file as well. So now when we update the starting html-file to redirect to a new version we will get a new panel content but old content for the custom-functions, although the html-file will contain a link to a new custom-functions file. Seems like a link from the manifest takes precedence over the link in the html-file. That's why we still need to have manifest updates for any custom-functions changes.
We'd rather not update the custom functions file in the same place because it's good to have consistency over panel content and functions version. Is there any way we can load and update all content from a single starting point?

**Comment by jargil** [comment]
@AlexPhilippov - very useful details, thank you! Some of the things you mentioned fall outside my area of expertise so I've asked one of my peers to chime in.

**Comment by madhavagrawal17** [comment]
@AlexPhilippov, Thank you for your explanation. I wanted to provide some clarification on the design for SharedRuntime. 
1. If you are using SharedRuntime then the Custom functions JS file will be part of the HTML page and it will use the latest JavaScript as we don’t cache or use the JavaScript provided in the manifest for the SharedRuntime.
2. We only use the Metadata file (JSON file) from the manifest and cache it. This will affect the update if you change the Functions definitions or Functions properties. In this case, the update might take a couple of days to fetch the latest JSON metadata file in order to update the cache and install the new functions.

I also want to know for which above points (1. or 2.) you are facing the update problem?

**Comment by AlexPhilippov** [comment]
@madhavagrawal17, we expect that metadata update might require additional efforts. For now, we struggling with point 1. Invalidating cache for HTML-file with the reference to the new functions JS-file won't give us updated content in several hours.
Regarding point 2, new manifest approval won't give us updates of the content as well, we have to wait for several days to get the new content. We recently require updates for the metadata file. Manifest updates were submitted just for forcing the content update.
Another detail that might be affecting the whole update process. We have a Start.html file referenced in manifest that redirects to the current version of the panel content with js-code. It now contains a link to functions JS-file after xiaochun's comment. So each version update we're invalidating Start.html with a new link to functions file and js-code redirecting to a new panel content HTML-file with the same link. To be clear, now Start.html contains the link to /3.3.1.4/Js/functions.js and redirects to /3.3.1.4/Views/Excel.html. Question is, after the Start.html invalidation, will Excel check the original Start.html file that's referenced in the manifest? Or it will be working with /3.3.1.4/Views/Excel.html until it detects some update for the manifest?

**Comment by madhavagrawal17** [comment]
@AlexPhilippov, I checked the details of your manifest and start.html file.  Firstly, to answer your question, Excel will always reference the start.html file specified in the manifest. 
Now let me point towards some issues that I see: 
1. You have used the start.html as the starting point so as to keep the URL in the manifest file the same and on Office.onReady() you redirect it to a specific version of panel content HTML file. If my understanding is correct you don't need the reference of functions.js in the start.html file as it is not required. According to your design if the redirection to the specific version of Excel.html is correct then the page will fetch the correct version of the functions.js file.
2. I think the main problem is with the Invalidation of the start.html which is not working. My suspicion is S3 is caching the start.html file and even after you update it and invalidate it, S3 is still serving the old file. You can test this by using the URL of start.html on Internet Explorer or Edge to check the content of your start.html after your update.
3. In order to update the JSON file you don't need Manifest update you can just update the content of the JSON file. We update the cache for the JSON file in a couple of days if the content of the JSON is updated based on the Etag. If a couple of days doesn't work for you then you need to ask your user to reinstall the Add-in or click on the Refresh button using instructions [here](https://support.microsoft.com/en-us/office/find-my-office-add-in-b022dc72-8ffc-4bfc-b1ff-c3f9a2182964)

**Comment by AlexPhilippov** [comment]
@madhavagrawal17, some comments on mentioned issues.
1. It was xiaochun's suggestion to add the link to functions.js. Without it, the 2nd case in the original description will still be reproducible. Please, check.
2. I checked the invalidation of HTML-file on S3 in IE and it seems to be working fine. To be clear I didn't just rewrite the file. There is a special command "CloudFront Invalidation" in my CloudBerry Explorer for S3. It requires special permissions to the folder. And this sends a new ETag and Last-Modified headers for the file.
3. We want to have consistency in each version. So, the metadata JSON file is stored as a part of the version. For now, adding changes to metadata seems rather uncommon and it's reasonable to update the manifest to a new version for such changes.

**Comment by madhavagrawal17** [comment]
@AlexPhilippov, find my reply to your comments.

1. I think adding functions.js in start.html won’t solve any of your problems.  I think uninstalling and reinstalling the add-in doesn’t install custom functions for you is a different problem. I think Xiaochun suggested that change because he was thinking start.html is your final html, but it is not the case as per your design.
2. I tested your scenario with AWS S3 account with “CloudFront Invalidation”, but it seems to be working for me. I used the same design by creating a sharedruntime.html which is redirecting to content.html. In my test scenario I am invalidating the sharedruntime.html and it works perfectly for me.
3. If you think adding changes to metadata is uncommon for your sceanrio and updating the manifest is reasonable for you then I will leave that up to you.
